### ISSUE TITLE
Database Security Audit

### ISSUE THREAD

**Issue by jhigginbotham64** [Vize-Programming-Team/vize-meteor-app#100]
The app currently connect to Amazon RDS with a username and password that gives it *_admin_* access to the database. This might be fine and good if we could guarantee that that connection object (defined in /imports/api/graphql/connectors/postgresql.js) was completely local to the server, since we already know that credentials used are read from environment variables which are local to the server, but I'm not convinced. _Is there anything preventing a malicious custom-made client from importing (or otherwise gaining access to) the connection object and doing whatever they want with it?_
Further, _should we even be letting the website use the admin account?_ If not, what exactly are the permissions that it does need?
Finally, are there other database security best-practices that I may have overlooked while setting up the website? 
I'm requesting help from the rest of you because I may not be around much longer, and this isn't something I feel like I have the time and freedom to work on right now, given that I have no idea how serious the vulnerabilities are or whether they even exist.

**Comment by jhigginbotham64** [comment]
I said "admin" access, but that may not be technically correct: the account used by app is just the original one created by RDS on database initialization, and it uses RDS's defaults configuration options, which may not necessarily add up to "admin". AFAIK no one has true "admin" access to the database except the AWS IAM user all the necessary permissions. 
Which just goes to show that we need to learn more about this...

**Comment by joshuatshaffer** [comment]
> Further, should we even be letting the website use the admin account? If not, what exactly are the permissions that it does need?
No, it should only have read/write access to the data. It does not need to reconfigure the database. As I understand it all of that work is done in the init scripts.

**Comment by joshuatshaffer** [comment]
> Finally, are there other database vulnerabilities that I haven't noticed? Or security best-practices that I may have overlooked while setting up the backend?

We really should set up some kind of [IAC](https://en.wikipedia.org/wiki/Infrastructure_as_Code) for this (I'm thinking [Terraform](https://terraform.io)). It can generate and deploy accounts and passwords automatically.

Additionally, we could set up a firewall/VPC on the database to allow only connections from the `galaxy.meteor.com` servers.

**Comment by joshuatshaffer** [comment]
> Is there anything preventing a malicious custom-made client from importing (or otherwise gaining access to) the connection object and doing whatever they want with it?

Not that I know of. :sweat_smile:

In theory the web/api server only gives access via the GraphQL API which enforces the business rules, **BUT** if someone somehow got the username and password for the database... we are **totally screwed**. The way it is setup right now, anyone on any computer with an internet connection, a PostgreSQL client, the username and password, and the database's ip/hostname can get full admin access to the database.

We can reduce this danger with some best practices like a firewall, VPC, non-admin accounts, etc. We will never be 100% safe, but the more layers of security the less likely we will get hacked.

**Comment by jhigginbotham64** [comment]
> > Is there anything preventing a malicious custom-made client from importing (or otherwise gaining access to) the connection object and doing whatever they want with it?
> 
> Not that I know of. sweat_smile
> 
> In theory the web/api server only gives access via the GraphQL API which enforces the business rules, **BUT** if someone somehow got the username and password for the database... we are **totally screwed**. The way it is setup right now, anyone on any computer with an internet connection, a PostgreSQL client, the username and password, and the database's ip/hostname can get full admin access to the database.
> 
> We can reduce this danger with some best practices like a firewall, VPC, non-admin accounts, etc. We will never be 100% safe, but the more layers of security the less likely we will get hacked.

Okay...so far I've been assuming that the username and password can be (and are) kept safe, i.e. in the environment variables on the server. What I'm more concerned with is whether the connection pool and wrappers exported in imports/api/graphql/connectors/postgresql.js **_are ever exposed in a way that would make them vulnerable_**, because anyone who uses them once they've been initialized doesn't need to know (or need to have known) the database username and password.

This is not to say that I'm not interested in the suggestions you had for protecting the database against malicious users who obtain the credentials, but my question was most directly concerned with **_how much of an attack surface is presented by the code in imports/api/data and the code that uses exports from that directory_**.

**Comment by joshuatshaffer** [comment]
I have just found out that we can use "roles" to set up a separate login with lower permissions for the API server. This way we can restrict the server's access to the database so that even if the API is compromised the attackers will still not have full admin access to the database. https://www.postgresql.org/docs/9.6/static/user-manag.html

**Comment by jhigginbotham64** [comment]
> > I have just found out that we can use "roles" to set up a separate login with lower permissions for the API server. This way we can restrict the server's access to the database so that even if the API is compromised the attackers will still not have full admin access to the database. https://www.postgresql.org/docs/9.6/static/user-manag.html
> 
> Yep, I was aware of this. It's just a matter of setting aside time, figuring it out, and actually doing it. Although it's interesting, now that you mention it, to think that my earlier scenario seems to be mostly equivalent to the API being abused, which is something else we may still want to think about.

Actually, since this is now a public issue on GitHub, and I already figured out some of what I'll need to know while working with Charter, I'm going to stop what I'm doing and move this to the top of my Vize to-do list...
